### PR TITLE
[python] Consolidate all `_set_coords` methods into single `_util` function

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**

As part of work for https://github.com/single-cell-data/TileDB-SOMA/issues/3053.

**Changes:**

This is clean-up work.

This takes the separate implementations of `_set_coords` from `_dataframe`, `_sparse_nd_array`,  `_spatial_dataframe`, and `_soma_array` into a single `_util._set_coords`. `_cast_domainish` was also moved out of `SOMAArrayWrapper` and into `_util` to support casting of domains and NEDs in `_set_coord`.

**Note for reviewers:**

I am unsure if `_util.py` is the appropriate place for this.

### CODE REVIEW

File: apis/python/src/tiledbsoma/_dataframe.py [hunk 1]
```diff
@@ -696,7 +693,7 @@ def read(
         if value_filter is not None:
             sr.set_condition(QueryCondition(value_filter), handle.schema)
 
-        self._set_coords(sr, coords)
+        _util._set_coords(sr, coords)
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/_dataframe.py:696]: Maybe not on this PR (or maybe on this PR!) can we rename `sr` (which I think used to stand for _soma reader_) into something more reflective of what it currently is? Maybe as simple as `sarr` ......

**Review by johnkerl — APPROVED** [review]
This is beautiful ... thank you!!

**Comment by johnkerl** [comment]
> I am unsure if `_util.py` is the appropriate place for this.

@nguyenv I think this is fine :)

File: apis/python/src/tiledbsoma/_dataframe.py (hunk 1 above)
Comment (nguyenv) [reply apis/python/src/tiledbsoma/_dataframe.py:696]: Yeah. I'll do that on this PR.

**Pull request merged**